src/Bicep.Decompiler/Rewriters/ForExpressionSimplifierRewriter.cs

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using Bicep.Core.Syntax.Rewriters;
+using Bicep.Core.Syntax.Visitors;
+
+namespace Bicep.Core.Decompiler.Rewriters
+{
+    /// <summary>
+    /// Looks for syntax of form (commonly generated by the decompiler):
+    ///   for {index} in range(0, length({array}))
+    /// Where elements are being accessed via:
+    ///   {array}[{index}]
+    ///
+    /// Rewrites this syntax instead to:
+    ///   for {item} in {array}
+    /// Where elements are accessed via:
+    ///   {item}
+    /// </summary>
+    public class ForExpressionSimplifierRewriter : SyntaxRewriteVisitor
+    {
+        private readonly SemanticModel semanticModel;
+
+        public ForExpressionSimplifierRewriter(SemanticModel semanticModel)
+        {
+            this.semanticModel = semanticModel;
+        }
+
+        private static bool IsLoopIteratorExpression(SyntaxBase syntax, [NotNullWhen(true)] out VariableAccessSyntax? arrayVariable)
+        {
+            arrayVariable = null;
+
+            if (syntax is not FunctionCallSyntax rangeFunction ||
+                !LanguageConstants.IdentifierComparer.Equals(rangeFunction.Name.IdentifierName, "range") ||
```

Comment:
>range [](start = 97, length = 5)

Shouldn't this check request symbol for the syntax node and then check if it's a FunctionSymbol with name "range"? (Could be encapsulated in a helper method.) This would remove assumptions about binding logic from this check. 

(Same comment for the similar pattern below as well.)

Reply:
Pushed some changes - take a look!

Reply:
I like it. Worth encapsulating the name check into another helper also? I ran into a similar problem elsewhere and could use it as well. (Totally optional.)

---
In reply to: [594717251](https://github.com/Azure/bicep/pull/1873#discussion_r594717251) [](ancestors = 594717251)